manager/encryption/encryption.go

```diff
@@ -0,0 +1,109 @@
+package encryption
+
+import (
+	"crypto/rand"
+	"fmt"
+
+	"github.com/docker/swarmkit/api"
+	"github.com/gogo/protobuf/proto"
+)
+
+// the size of any secret data generated by this package
+const secretSize = 32
+
+// This package defines the interfaces and encryption package
+
+// ErrCannotDecode is the type of error returned when some data cannot be decoded as plaintext
+type ErrCannotDecode struct {
+	msg string
+}
+
+func (e ErrCannotDecode) Error() string {
+	return e.msg
+}
+
+// A Decoder can decrypt an encrypted record
+type Decoder interface {
+	Decode(api.EncryptedRecord) ([]byte, error)
+}
+
+// A Encoder can encrypt some bytes into an encrypted record
+type Encoder interface {
+	Encode(data []byte) (*api.EncryptedRecord, error)
+}
+
+type noopCoder struct{}
+
+func (n noopCoder) Decode(e api.EncryptedRecord) ([]byte, error) {
+	if e.Algorithm != n.Algorithm() {
+		return nil, fmt.Errorf("not an unencrypted record")
+	}
+	return e.Data, nil
+}
+
+func (n noopCoder) Encode(data []byte) (*api.EncryptedRecord, error) {
+	return &api.EncryptedRecord{
+		Algorithm: n.Algorithm(),
+		Data:      data,
+	}, nil
+}
+
+func (n noopCoder) Algorithm() api.EncryptedRecord_Algorithm {
+	return api.NotEncrypted
+}
+
+// NoopCoder is just a pass-through coder - it does not actually encode or
+// decode any data
+var NoopCoder = noopCoder{}
+
+// Decode turns a slice of bytes serialized as an EncryptedRecord into a slice of plaintext bytes
+func Decode(encoded []byte, decoder Decoder) ([]byte, error) {
+	if decoder == nil {
+		decoder = NoopCoder
+	}
+	r := api.EncryptedRecord{}
+	if err := proto.Unmarshal(encoded, &r); err != nil || r.Size() != len(encoded) {
```

Comment:
Do you mean unmarshalling the WAL when reading, and then unmarshalling the encrypted data inside the WAL?  I'm not sure - I haven't done a performance test.  But that's also the only way we can inject the extra encryption information when we do the encryption at this layer (when writing raft data to disk), unless we want to reimplement etcd's wal and snapshot packages instead of wrapping them.